src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetWebApi2Integration.cs

```diff
@@ -24,36 +26,46 @@ public static class AspNetWebApi2Integration
         /// <returns>A task with the result</returns>
         public static object ExecuteAsync(object @this, object controllerContext, object cancellationTokenSource)
+        {
+            var cancellationToken = ((CancellationTokenSource)cancellationTokenSource).Token;
+            return ExecuteAsync(@this, controllerContext, cancellationToken);
+        }
+
+        /// <summary>
+        /// ExecuteAsync calls the underlying ExecuteAsync and traces the request.
+        /// </summary>
+        /// <param name="this">The Api Controller</param>
+        /// <param name="controllerContext">The controller context for the call</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>A task with the result</returns>
+        public static async Task<HttpResponseMessage> ExecuteAsync(object @this, object controllerContext, CancellationToken cancellationToken)
         {
             Type controllerType = @this.GetType();
 
             // in some cases, ExecuteAsync() is an explicit interface implementation,
             // which is not public and has a different name, so try both
             var executeAsyncFunc =
-                DynamicMethodBuilder<Func<object, object, CancellationToken, object>>
+                DynamicMethodBuilder<Func<object, object, CancellationToken, Task<HttpResponseMessage>>>
                    .GetOrCreateMethodCallDelegate(controllerType, "ExecuteAsync") ??
-                DynamicMethodBuilder<Func<object, object, CancellationToken, object>>
+                DynamicMethodBuilder<Func<object, object, CancellationToken, Task<HttpResponseMessage>>>
                    .GetOrCreateMethodCallDelegate(controllerType, "System.Web.Http.Controllers.IHttpController.ExecuteAsync");
 
             using (Scope scope = CreateScope(controllerContext))
             {
-                return scope.Span.Trace(
-                                        () =>
-                                        {
-                                            CancellationToken cancellationToken = ((CancellationTokenSource)cancellationTokenSource).Token;
-                                            return executeAsyncFunc(@this, controllerContext, cancellationToken);
-                                        },
-                                        onComplete: e =>
-                                        {
-                                            if (e != null)
-                                            {
-                                                scope.Span.SetException(e);
-                                            }
-
-                                            // some fields aren't set till after execution, so repopulate anything missing
-                                            UpdateSpan(controllerContext, scope.Span);
-                                            scope.Span.Finish();
-                                        });
+                try
```

Comment:
What prevents using `scope.Span.Trace` here?

Reply:
Using `Task.ContinueWith()` wraps exceptions inside an `AggregateException`, changing the behavior of the profiled application. This PR fixes an issue that Kibo is having where they handle a specific exception type and return a 404 http status code, which the front end is expecting. When they enable our tracer, the exception type is wrong and their app returns a 500 instead.

Using `async`/`await` results in cleaner code and does not wrap the exception thrown.

Reply:
Changing extension method `Span.Trace()` to use `async`/`await` was only temporary. I was going to remove `Span.Trace()` entirely in a separate PR, but I just went ahead and did it here instead (see new commits).

That method was too complicated with little benefit. It checked for return types and had different behavior for `Task`, `Task<T>`, and anything else. It also had another problem: if the result was a `Task` or `Task<T>`, the `Scope` could be disposed before the task was completed.